scripts/o3de/o3de/repo.py

```diff
@@ -106,27 +113,65 @@ def validate_remote_repo(repo_uri: str, validate_contained_objects: bool = False
-        for template_json in template_set:
-            if not validation.valid_o3de_template_json(template_json):
-                logger.error(f'Invalid template JSON - {template_json} could not be loaded or is missing required values')
+        if repo_schema_version == REPO_IMPLICIT_SCHEMA_VERSION:
+            if download_object_manifests(repo_data) != 0:
                 return False
+            gem_set = get_gem_json_paths_from_cached_repo(repo_uri)
+            for gem_json in gem_set:
+                if not validation.valid_o3de_gem_json(gem_json):
+                    logger.error(f'Invalid gem JSON - {gem_json} could not be loaded or is missing required values')
+                    return False
+            project_set = get_project_json_paths_from_cached_repo(repo_uri)
+            for project_json in project_set:
+                if not validation.valid_o3de_project_json(project_json):
+                    logger.error(f'Invalid project JSON - {project_json} could not be loaded or is missing required values')
+                    return False
+            template_set = get_template_json_paths_from_cached_repo(repo_uri)
+            for template_json in template_set:
+                if not validation.valid_o3de_template_json(template_json):
+                    logger.error(f'Invalid template JSON - {template_json} could not be loaded or is missing required values')
+                    return False
+                
+        elif repo_schema_version == REPO_SCHEMA_VERSION_1_0_0:
+            gem_list = repo_data.get("gems", [])
+            for gem_json in gem_list:
+                if not validation.valid_o3de_gem_json_data(gem_json):
+                    logger.error(f'Invalid gem JSON - {gem_json} is missing required values')
+                    return False
+
+                # validate versioning info
+                if "versions_data" in gem_json:
+                    versions_data = gem_json["versions_data"]
+                    for version in versions_data:
+                        if not all(key in version for key in ['last_updated', 'version']):
+                            return False
+                        
+                        source_control_uri_defined = "source_control_uri" in version
+                        
+                        download_source_uri_defined = "download_source_uri" in version
+                        origin_uri_defined = "origin_uri" in version
+                        
+                        # prevent mixing fields intended for backwards compatibility (using XOR to verify)
+                        download_origin_defined = ((download_source_uri_defined and not origin_uri_defined) or
+                                                   (not download_source_uri_defined and origin_uri_defined))
+
+                        if not (source_control_uri_defined or download_origin_defined):
+                            return False
```

Comment:
I'm tempted to say lets move this validation code into `validation.py` as `valid_o3de_versions_data`, but it's likely only going to be used here so I'm on the fence about it.  A future developer might appreciate all the data validation in one place, but they might also appreciate the locality of the logic in one place to. Thoughts?

Reply:
Also, this section does not have any `logger.error` code in case it fails.  We should definitely have some thing state 'Invalid versions_data" so developers know where to look at least

Reply:
I had the same exact split in thoughts. The only thing preventing me from moving it outright to validation is that our introduction of versioning for repos is very new. Plus there's a chance right now our schema plans will evolve, so it feels best to keep the logic localized for now. If we do get it ironed out more in the future, I would be okay moving it to validation at that time.